lib/puppet/util/postgresql_validator.rb

```diff
@@ -0,0 +1,64 @@
+module Puppet
+  module Util
+    class PostgresqlValidator
+      attr_reader :resource
+
+      def initialize(resource)
+        @resource = resource
+      end
+
+      def build_psql_cmd
```

Comment:
Postgres has the `PQping()` command on its binary protocol to issue a connection check.  It can be used with `pg_isready` binary.  I think it would make a good default provider.  It would be more efficient and easier to use, because you don't need to pass authentication.  It also supports a custom timeout.